src/index.ts

```diff
@@ -0,0 +1,129 @@
+type MoneyField = {
+    type: 'money';
+    currency_code: CurrencyCode;
+} & Field;
+
+type NumberField = {
+    type: 'number';
+} & Field;
+
+type PercentageField = {
+    type: 'percentage';
+} & Field;
+
+type StringField = {
+    type: 'string';
+} & Field;
+
+type Fields = Record<string, DateField | DateTimeField | DurationField | MoneyField | NumberField | PercentageField | StringField>
+
+type KeysMatching<T, V> = { [K in keyof T]-?: T[K] extends V ? K : never }[keyof T];
+
+type Schema<T extends Fields> = {
+    id: string;
+    fields: T;
+    uniqueBy: Array<KeysMatching<T, StringField | DateField>>;
+}
+
+type FieldType<T> = T extends DateField
+    ? string
+    : T extends DateTimeField
+    ? string
+    : T extends DurationField
+    ? number
+    : T extends MoneyField
+    ? number
+    : T extends NumberField
+    ? number
+    : T extends PercentageField
+    ? number
+    : T extends StringField
+    ? string
+    : never;
+
+type IsOptional<K extends keyof F, F extends Fields> = F[K] extends OptionalField ? K : never;
+type IsRequired<K extends keyof F, F extends Fields> = F[K] extends OptionalField ? never : K;
+
+type DatasetDataItem<F extends Fields> = {
+    [K in keyof F as IsRequired<K, F>]: FieldType<F[K]>;
+} & {
+        [K in keyof F as IsOptional<K, F>]?: FieldType<F[K]>;
+    }
+
+class Dataset<T extends Fields> {
+    // can we get these directly from the schema?
+    id: string;
+    fields: T;
+    uniqueBy: Array<KeysMatching<T, StringField | DateField>>;
+
+    constructor(schema: Schema<T>) {
+        this.id = schema.id;
+        this.fields = schema.fields;
+        this.uniqueBy = schema.uniqueBy;
+    }
+
+    create(): Promise<Schema<T>> {
+        const { id, fields, uniqueBy } = this;
+        return Promise.resolve({ id, fields, uniqueBy });
+    }
+
+    post(items: DatasetDataItem<T>[], deleteBy?: keyof T): Promise<void> {
```

Comment:
what do you think about calling these append and replace rather than post and put?

Reply:
I'd be happy to make that change 👍